src/ledger.js

```diff
@@ -0,0 +1,131 @@
+// LedgerSystem class
+const BigNumber = require('bignumber.js');
+const Data = require("./database");
+const FIFOQueue = require("./queue");
+const sleep = require('./utils').sleep;
+const groupBy = require('./utils').groupBy;
+const OperationTypes = {
+    TRANSFER: "TRANSFER"
+}
+const OperationStatus = {
+    INIT: "INIT",
+    PROCESSING: "PROCESSING",
+    APPLIED: "APPLIED",
+    REJECTED: "REJECTED"
+}
+const Entities = {
+    OPERATIONS: "operations",
+    ACCOUNTS: "accounts",
+    ENTRIES: "entries"
+}
+class LedgerSystem {
+    // singleton class
+    constructor (databaseConfig) {
+        this.db = new Data[databaseConfig.class](databaseConfig);
+        this.postingQueue = new FIFOQueue(this.db.getAll(Entities.OPERATIONS).filter(operation => [OperationStatus.INIT, OperationStatus.PROCESSING].includes(operation.status)).sort((a, b) => a.id - b.id).map((operation) => {
+            return async () => {
+                await this.postOperationEntries(operation.id, operation.entries)
+            }
+        }));
+    }
+    validateEntries(entries) {
+        const accountGroupedEntries = groupBy(entries, "accountId");
+        Object.keys(accountGroupedEntries).forEach((accountId) => {
+            const account = this.db.get(Entities.ACCOUNTS, accountId);
+            if (account && account.restrictions) {
+                const accountEntries = accountGroupedEntries[accountId];
+                const accountBalances = this.getAccountBalances(accountId);
+                const assetGroupedEntries = groupBy(accountEntries, "assetId");
+                Object.keys(assetGroupedEntries).forEach((assetId) => {
+                    const assetBalance = accountBalances[assetId] || "0";
+                    let newEntriesSum
+                    if (assetGroupedEntries[assetId].length > 1) {
+                        newEntriesSum = assetGroupedEntries[assetId].reduce((r, i) => (BigNumber(r.value || r).plus(BigNumber(i.value))).toString());
+                    } else {
+                        newEntriesSum = assetGroupedEntries[assetId][0].value;
+                    }
+                    if (account.restrictions && account.restrictions.minimumCreditBalance && (BigNumber(newEntriesSum).plus(BigNumber(assetBalance))).isLessThan(BigNumber(account.restrictions.minimumCreditBalance))) {
```

Comment:
this is the one and only place the word 'Credit' is used in this system. think long and hard if you need it before you add new words to the domain's language.

Reply:
as we replaced Credit with +ve values and Debit with -ve values, I renamed the key to `minBalance`.